File: diagram/class_diagrams/puml_to_drawio.py
from __future__ import annotations

import argparse
import math
import os
import re
import sys
import xml.etree.ElementTree as ET
from dataclasses import dataclass
from pathlib import Path
from typing import Dict, List, Optional, Tuple


@dataclass
class UmlBox:
    kind: str  # 'class' | 'enum' | other
    name: str
    body_lines: List[str]


@dataclass
class UmlEdge:
    src: str
    dst: str
    arrow: str
    label: str


CLASS_BLOCK_RE = re.compile(r"^(class|interface|abstract\s+class)\s+([A-Za-z0-9_$.]+)\s*\{\s*$")
ENUM_BLOCK_RE = re.compile(r"^enum\s+([A-Za-z0-9_$.]+)\s*\{\s*$")
REL_RE = re.compile(
    r"^([A-Za-z0-9_$.]+)\s+([.\-*o<|/\\]+--+>?|--+>?|\.\.+>?|<\|--|\*--|o--|\+--|--\|>|-->?|<-->)\s+([A-Za-z0-9_$.]+)\s*(?::\s*(.*))?$"
)


def _clean_member_line(line: str) -> str:
    return line.rstrip("\n").rstrip("\r")


def parse_puml(text: str) -> Tuple[List[UmlBox], List[UmlEdge], str]:
    boxes: List[UmlBox] = []
    edges: List[UmlEdge] = []

    title = ""
    lines = text.splitlines()

    i = 0
    while i < len(lines):
        raw = lines[i]
        line = raw.strip()

        if line.lower().startswith("title "):
            title = raw.strip()[6:].strip()
            i += 1
            continue

        if not line or line.startswith("'"):
            i += 1
            continue

        if line.startswith("@") or line.startswith("skinparam") or line.startswith("hide "):
            i += 1
            continue

        m_class = CLASS_BLOCK_RE.match(line)
        if m_class:
            kind = "class"
            name = m_class.group(2)
            body: List[str] = []
            i += 1
            while i < len(lines):
                inner = lines[i]
                if inner.strip() == "}":
                    break
                inner_stripped = inner.rstrip("\n").rstrip("\r")
                if inner_stripped.strip() != "":
                    body.append(_clean_member_line(inner_stripped.strip()))
                i += 1
            boxes.append(UmlBox(kind=kind, name=name, body_lines=body))
            i += 1
            continue

        m_enum = ENUM_BLOCK_RE.match(line)
        if m_enum:
            kind = "enum"
            name = m_enum.group(1)
            body: List[str] = []
            i += 1
            while i < len(lines):
                inner = lines[i]
                if inner.strip() == "}":
                    break
                inner_stripped = inner.rstrip("\n").rstrip("\r")
                if inner_stripped.strip() != "":
                    body.append(_clean_member_line(inner_stripped.strip()))
                i += 1
            boxes.append(UmlBox(kind=kind, name=name, body_lines=body))
            i += 1
            continue

        m_rel = REL_RE.match(line)
        if m_rel:
            src, arrow, dst, label = m_rel.group(1), m_rel.group(2), m_rel.group(3), m_rel.group(4) or ""
            # Normalize arrow a bit; we don't need exact semantics for draw.io rendering.
            edges.append(UmlEdge(src=src, dst=dst, arrow=arrow, label=label.strip()))
            i += 1
            continue

        i += 1

    return boxes, edges, title


def xml_escape(s: str) -> str:
    return (
        s.replace("&", "&amp;")
        .replace("<", "&lt;")
        .replace(">", "&gt;")
        .replace('"', "&quot;")
        .replace("'", "&apos;")
    )


def to_drawio_value(box: UmlBox) -> str:
    header = f"<div><b>{xml_escape(box.name)}</b>"
    if box.kind == "enum":
        header += " <i>&lt;&lt;enum&gt;&gt;</i>"
    header += "</div>"

    if not box.body_lines:
        return header

    body = "<div style=\"margin-top:4px;\">" + "<br/>".join(xml_escape(x) for x in box.body_lines) + "</div>"
    return header + body


def estimate_size(box: UmlBox) -> Tuple[int, int]:
    # Very rough sizing to avoid all boxes overlapping.
    lines = [box.name] + box.body_lines
    max_len = max((len(l) for l in lines), default=len(box.name))
    line_count = max(1, len(lines))

    char_w = 6.5
    line_h = 16
    padding_w = 40
    padding_h = 20

    w = int(min(520, max(180, math.ceil(max_len * char_w) + padding_w)))
    h = int(min(420, max(60, line_count * line_h + padding_h)))
    return w, h


def build_drawio(boxes: List[UmlBox], edges: List[UmlEdge], title: str) -> str:
    # IDs
    next_id = 2
    cell_id_by_name: Dict[str, str] = {}

    mxfile = ET.Element("mxfile", attrib={
        "host": "app.diagrams.net",
        "modified": "2026-03-23T00:00:00.000Z",
        "agent": "puml_to_drawio.py",
        "version": "24.7.10",
        "type": "device",
    })

    diagram = ET.SubElement(mxfile, "diagram", attrib={"id": "puml", "name": title or "Page-1"})

    model = ET.SubElement(diagram, "mxGraphModel", attrib={
        "dx": "1200",
        "dy": "800",
        "grid": "1",
        "gridSize": "10",
        "guides": "1",
        "tooltips": "1",
        "connect": "1",
        "arrows": "1",
        "fold": "1",
        "page": "1",
        "pageScale": "1",
        "pageWidth": "1169",
        "pageHeight": "827",
        "math": "0",
        "shadow": "0",
    })

    root = ET.SubElement(model, "root")
    ET.SubElement(root, "mxCell", attrib={"id": "0"})
    ET.SubElement(root, "mxCell", attrib={"id": "1", "parent": "0"})

    # Simple grid layout
    cols = 3
    x0, y0 = 40, 40
    x_gap, y_gap = 40, 40

    for idx, box in enumerate(boxes):
        w, h = estimate_size(box)
        col = idx % cols
        row = idx // cols
        x = x0 + col * (520 + x_gap)
        y = y0 + row * (420 + y_gap)

        cid = str(next_id)
        next_id += 1
        cell_id_by_name[box.name] = cid

        style = "rounded=0;whiteSpace=wrap;html=1;align=left;verticalAlign=top;spacingLeft=8;spacingTop=6;"
        if box.kind == "enum":
            style += "dashed=1;"

        cell = ET.SubElement(root, "mxCell", attrib={
            "id": cid,
            "value": to_drawio_value(box),
            "style": style,
            "vertex": "1",
            "parent": "1",
        })
        ET.SubElement(cell, "mxGeometry", attrib={
            "x": str(x),
            "y": str(y),
            "width": str(w),
            "height": str(h),
            "as": "geometry",
        })

    # Add edges
    for e in edges:
        if e.src not in cell_id_by_name or e.dst not in cell_id_by_name:
            continue
        eid = str(next_id)
        next_id += 1

        style = "endArrow=block;endFill=1;edgeStyle=orthogonalEdgeStyle;rounded=0;html=1;"
        if ".." in e.arrow:
            style += "dashed=1;"

        edge_cell = ET.SubElement(root, "mxCell", attrib={
            "id": eid,
            "value": xml_escape(e.label) if e.label else "",
            "style": style,
            "edge": "1",
            "parent": "1",
            "source": cell_id_by_name[e.src],
            "target": cell_id_by_name[e.dst],
        })
        ET.SubElement(edge_cell, "mxGeometry", attrib={"relative": "1", "as": "geometry"})

    # Serialize
    xml_str = ET.tostring(mxfile, encoding="utf-8")
    return xml_str.decode("utf-8")


def convert_file(src_path: Path, out_path: Path) -> None:
    text = src_path.read_text(encoding="utf-8")
    boxes, edges, title = parse_puml(text)
    if not boxes:
        raise RuntimeError(f"No classes/enums found in {src_path}")
    xml = build_drawio(boxes, edges, title=title)
    out_path.write_text(xml, encoding="utf-8")


def main(argv: Optional[List[str]] = None) -> int:
    p = argparse.ArgumentParser(description="Convert simple PlantUML class diagrams (.puml) to draw.io (.drawio) XML")
    p.add_argument("inputs", nargs="+", help="Input .puml files or directories")
    p.add_argument("--out-dir", default="", help="Output directory (defaults to alongside input files)")
    p.add_argument("--overwrite", action="store_true", help="Overwrite existing .drawio files")

    args = p.parse_args(argv)

    input_paths: List[Path] = [Path(x) for x in args.inputs]
    out_dir = Path(args.out_dir) if args.out_dir else None

    puml_files: List[Path] = []
    for ip in input_paths:
        if ip.is_dir():
            puml_files.extend(sorted(ip.glob("**/*.puml")))
        else:
            puml_files.append(ip)

    if not puml_files:
        print("No .puml files found", file=sys.stderr)
        return 2

    converted = 0
    for src in puml_files:
        if src.suffix.lower() != ".puml":
            continue
        target_dir = out_dir if out_dir else src.parent
        target_dir.mkdir(parents=True, exist_ok=True)
        out = target_dir / (src.stem + ".drawio")
        if out.exists() and not args.overwrite:
            continue
        convert_file(src, out)
        converted += 1

    print(f"Converted {converted} file(s)")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: diagram/class_diagrams/puml_to_drawio.mjs
import * as fs from 'node:fs';
import * as path from 'node:path';
import * as crypto from 'node:crypto';

function htmlEscape(s) {
  return String(s)
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/\"/g, '&quot;')
    .replace(/'/g, '&#x27;');
}

function xmlAttrEscape(s) {
  return String(s)
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/\"/g, '&quot;')
    .replace(/'/g, '&apos;');
}

function clamp(min, v, max) {
  return Math.max(min, Math.min(max, v));
}

function uuid12() {
  return crypto.randomBytes(6).toString('hex');
}

/**
 * @typedef {{source:string, token:string, target:string, hidden:boolean, directionHint:('left'|'right'|'up'|'down'|null), label:string}} Relation
 * @typedef {{name:string, kind:'class'|'enum', bodyLines:string[], order:number, width:number, height:number, x:number, y:number}} Node
 */

const BLOCK_START_RE = /^\s*(class|enum|interface|abstract\s+class)\s+([A-Za-z_][\w.]*)\s*\{\s*$/;
const TITLE_RE = /^\s*title\s+(.+?)\s*$/;
const RANKSEP_RE = /^\s*skinparam\s+ranksep\s+(\d+)\s*$/;
const NODESEP_RE = /^\s*skinparam\s+nodesep\s+(\d+)\s*$/;
const RELATION_RE =
  /^\s*([A-Za-z_][\w.]*)\s+([-.o*|<>]+(?:\[[^\]]+\])?(?:left|right|up|down)?[-.<>|]*)\s+([A-Za-z_][\w.]*)\s*(?::\s*(.+))?$/;

function normalizeRelationToken(token) {
  let t = String(token);
  t = t.replace(/\[[^\]]+\]/g, '');
  t = t.replace(/left|right|up|down/g, '');
  t = t.trim();
  // Reduce very long tokens to core arrows when possible.
  const known = ['..>', '-->', '*--', 'o--', '--|>', '..|>'];
  for (const k of known) {
    if (t.includes(k)) return k;
  }
  return t;
}

function parsePumlFile(filePath) {
  const text = fs.readFileSync(filePath, 'utf8');
  const lines = text.split(/\r?\n/);

  let title = path.basename(filePath, '.puml');
  let ranksep = 120;
  let nodesep = 70;
  let direction = 'LR';

  /** @type {Map<string, Node>} */
  const nodes = new Map();
  /** @type {Relation[]} */
  const relations = [];

  let order = 0;
  for (let i = 0; i < lines.length; i++) {
    const raw = lines[i] ?? '';
    const stripped = raw.trim();

    if (stripped === 'left to right direction') {
      direction = 'LR';
      continue;
    }

    const mTitle = raw.match(TITLE_RE);
    if (mTitle) {
      title = mTitle[1];
      continue;
    }

    const mRank = raw.match(RANKSEP_RE);
    if (mRank) {
      ranksep = Number.parseInt(mRank[1], 10);
      continue;
    }

    const mNodeSep = raw.match(NODESEP_RE);
    if (mNodeSep) {
      nodesep = Number.parseInt(mNodeSep[1], 10);
      continue;
    }

    if (stripped.startsWith("'")) continue;
    if (stripped.startsWith('@') || stripped.startsWith('skinparam') || stripped.startsWith('hide ')) continue;

    const mBlock = raw.match(BLOCK_START_RE);
    if (mBlock) {
      const kindToken = mBlock[1];
      const name = mBlock[2];
      // Teacher requirement: enums are NOT shown in class diagrams.
      // Skip enum blocks entirely.
      if (kindToken.startsWith('enum')) {
        i++;
        while (i < lines.length && (lines[i] ?? '').trim() !== '}') i++;
        continue;
      }
      const kind = 'class';
      /** @type {string[]} */
      const bodyLines = [];
      i++;
      while (i < lines.length && (lines[i] ?? '').trim() !== '}') {
        bodyLines.push((lines[i] ?? '').replace(/\s+$/, ''));
        i++;
      }
      nodes.set(name, {
        name,
        kind,
        bodyLines,
        order,
        width: 0,
        height: 0,
        x: 0,
        y: 0,
      });
      order++;
      continue;
    }

    const mRel = raw.match(RELATION_RE);
    if (mRel) {
      const source = mRel[1];
      const token = mRel[2];
      const target = mRel[3];
      const label = (mRel[4] || '').trim();
      const hidden = token.includes('[hidden]');
      let directionHint = null;
      for (const candidate of ['left', 'right', 'up', 'down']) {
        if (token.includes(candidate)) {
          directionHint = candidate;
          break;
        }
      }
      relations.push({ source, token, target, hidden, directionHint, label });
      continue;
    }
  }

  return { title, ranksep, nodesep, direction, nodes, relations };
}

function splitSections(node) {
  const cleaned = (node.bodyLines || []).map((l) => String(l).trim()).filter((l) => l.length > 0);
  if (node.kind === 'enum') return { attrs: [], methodsOrValues: cleaned };

  const dividerIndex = cleaned.indexOf('--');
  if (dividerIndex >= 0) {
    return {
      attrs: cleaned.slice(0, dividerIndex),
      methodsOrValues: cleaned.slice(dividerIndex + 1),
    };
  }

  // Heuristic split (matches common PlantUML style where methods contain parentheses)
  // so draw.io can render a two-compartment class box like the reference.
  const attrs = [];
  const methodsOrValues = [];
  for (const line of cleaned) {
    // Keep explicit ellipsis lines but place them in the section they appear in.
    // (If the source has only one `...`, it will show in whichever section it belongs.)
    if (/\([^)]*\)/.test(line)) methodsOrValues.push(line);
    else attrs.push(line);
  }
  return { attrs, methodsOrValues };
}

function wrapText(line, width = 52) {
  const s = String(line ?? '');
  if (!s) return [''];

  /** @type {string[]} */
  const out = [];
  let current = '';
  const parts = s.split(/\s+/).filter(Boolean);

  const pushCurrent = () => {
    if (current) out.push(current);
    current = '';
  };

  for (const part of parts) {
    if (part.length > width) {
      pushCurrent();
      for (let i = 0; i < part.length; i += width) out.push(part.slice(i, i + width));
      continue;
    }

    if (!current) {
      current = part;
    } else if (current.length + 1 + part.length <= width) {
      current += ` ${part}`;
    } else {
      pushCurrent();
      current = part;
    }
  }
  pushCurrent();
  return out.length ? out : [s];
}

function buildNodeHtml(node) {
  const { attrs, methodsOrValues } = splitSections(node);

  /** @type {string[][]} */
  let sections = [];
  if (node.kind === 'enum') {
    // Enums: keep a single compartment of values.
    sections = [methodsOrValues];
  } else {
    // Teacher requirement: every class/interface must have 2 compartments:
    // top = attributes, bottom = methods (even if one side is empty).
    sections = [attrs, methodsOrValues];
  }

  /** @type {string[][]} */
  const wrappedSections = [];
  let maxChars = node.name.length;
  let totalLines = 0;

  for (const section of sections) {
    /** @type {string[]} */
    const wrappedLines = [];
    const effectiveLines = section.length > 0 ? section : [''];
    for (const line of effectiveLines) {
      for (const part of wrapText(line)) {
        wrappedLines.push(part);
        maxChars = Math.max(maxChars, part.length);
        totalLines += 1;
      }
    }
    wrappedSections.push(wrappedLines);
  }

  const width = clamp(140, maxChars * 6.2 + 28, 520);
  const headerHeight = 30;
  const lineHeight = 14;
  const sectionPadding = 8;
  const height = headerHeight + totalLines * lineHeight + sectionPadding * Math.max(1, wrappedSections.length);

  const rows = [];
  rows.push(
    `<tr><td style='text-align:center;font-weight:bold;padding:6px 4px;'>${htmlEscape(node.name)}</td></tr>`,
  );

  for (let index = 0; index < wrappedSections.length; index++) {
    const section = wrappedSections[index];
    const border = 'border-top:1px solid #000000;';

    // Section is never empty here because we omit empty compartments above,
    // but keep this defensive (and still avoid rendering a blank placeholder).
    const content =
      section.length > 0
        ? section
            .map((line) => {
              // Keep the compartment visible even when empty.
              const safe = line ? htmlEscape(line) : '&nbsp;';
              return `<div style='line-height:1.25;padding-left:2px;'>${safe}</div>`;
            })
            .join('')
        : '&nbsp;';
    rows.push(
      `<tr><td style='${border}vertical-align:top;padding:4px 4px 6px 4px;'>${content}</td></tr>`,
    );
  }

  const table = `<table style='width:100%;height:100%;border-collapse:collapse;'>${rows.join('')}</table>`;
  return { html: table, width, height };
}

function resolveLayout(nodes, relations, ranksep, nodesep, direction) {
  /** @type {Map<string, number>} */
  const cols = new Map();
  /** @type {Map<string, number>} */
  const rows = new Map();

  for (const name of nodes.keys()) {
    cols.set(name, 0);
    rows.set(name, 0);
  }

  for (let iter = 0; iter < nodes.size * 4; iter++) {
    let changed = false;
    for (const rel of relations) {
      if (!nodes.has(rel.source) || !nodes.has(rel.target)) continue;

      const srcCol = cols.get(rel.source) ?? 0;
      const dstCol = cols.get(rel.target) ?? 0;
      const srcRow = rows.get(rel.source) ?? 0;
      const dstRow = rows.get(rel.target) ?? 0;

      if (rel.hidden) {
        if (rel.directionHint === 'right' && dstCol < srcCol + 1) {
          cols.set(rel.target, srcCol + 1);
          changed = true;
        } else if (rel.directionHint === 'left' && srcCol < dstCol + 1) {
          cols.set(rel.source, dstCol + 1);
          changed = true;
        } else if (rel.directionHint === 'down' && dstRow < srcRow + 1) {
          rows.set(rel.target, srcRow + 1);
          changed = true;
        } else if (rel.directionHint === 'up' && srcRow < dstRow + 1) {
          rows.set(rel.source, dstRow + 1);
          changed = true;
        }
        continue;
      }

      if (direction === 'LR' && dstCol < srcCol + 1) {
        cols.set(rel.target, srcCol + 1);
        changed = true;
      }
    }
    if (!changed) break;
  }

  /** @type {Map<number, Node[]>} */
  const columns = new Map();
  for (const node of nodes.values()) {
    const col = cols.get(node.name) ?? 0;
    if (!columns.has(col)) columns.set(col, []);
    columns.get(col).push(node);
  }

  /** @type {Map<number, number>} */
  const xPositions = new Map();
  let xCursor = 40;
  const sortedCols = [...columns.keys()].sort((a, b) => a - b);
  for (const col of sortedCols) {
    const colNodes = columns.get(col) || [];
    const colWidth = Math.max(...colNodes.map((n) => n.width || 140));
    xPositions.set(col, xCursor);
    xCursor += colWidth + nodesep;
  }

  for (const col of sortedCols) {
    const colNodes = (columns.get(col) || []).slice();
    colNodes.sort((a, b) => {
      const ra = rows.get(a.name) ?? 0;
      const rb = rows.get(b.name) ?? 0;
      if (ra !== rb) return ra - rb;
      return a.order - b.order;
    });

    let yCursor = 70;
    let previousRow = 0;
    let previousHeight = 0;

    for (const node of colNodes) {
      const nodeRow = rows.get(node.name) ?? 0;
      const gapUnits = yCursor > 70 ? Math.max(1, nodeRow - previousRow) : Math.max(1, nodeRow + 1);
      if (yCursor === 70) {
        yCursor = 70 + nodeRow * Math.max(40, ranksep * 0.5);
      } else {
        yCursor += previousHeight + gapUnits * Math.max(40, ranksep * 0.45);
      }
      node.x = xPositions.get(col) ?? 40;
      node.y = yCursor;
      previousRow = nodeRow;
      previousHeight = node.height;
    }
  }
}

function makeEdgeStyle(token) {
  const base = [
    'edgeStyle=orthogonalEdgeStyle',
    'rounded=0',
    'orthogonalLoop=1',
    'jettySize=auto',
    'html=1',
    'strokeColor=#000000',
    'fontColor=#000000',
  ];

  const t = normalizeRelationToken(token);
  if (t === '..>') {
    base.push('dashed=1', 'dashPattern=8 8', 'endArrow=classic', 'endFill=1');
  } else if (t === '-->') {
    base.push('endArrow=classic', 'endFill=1');
  } else if (t === '*--') {
    base.push('startArrow=diamondThin', 'startFill=1', 'endArrow=none');
  } else if (t === 'o--') {
    base.push('startArrow=diamondThin', 'startFill=0', 'endArrow=none');
  } else if (t === '--|>') {
    base.push('endArrow=block', 'endFill=0');
  } else if (t === '..|>') {
    base.push('dashed=1', 'dashPattern=8 8', 'endArrow=block', 'endFill=0');
  } else {
    base.push('endArrow=none');
  }

  return `${base.join(';')};`;
}

function inferRelationLabel(token) {
  const t = normalizeRelationToken(token);
  if (t === '..>') return 'dependency';
  if (t === '-->' || t === '--') return 'association';
  if (t === '*--') return 'composition';
  if (t === 'o--') return 'aggregation';
  if (t === '--|>') return 'inheritance';
  if (t === '..|>') return 'realization';
  return '';
}

function buildDiagramXml(parsed) {
  const { title, ranksep, nodesep, direction, nodes, relations } = parsed;

  // First pass: compute node sizes.
  for (const node of nodes.values()) {
    const { width, height } = buildNodeHtml(node);
    node.width = width;
    node.height = height;
  }

  resolveLayout(nodes, relations, ranksep, nodesep, direction);

  /** @type {string[]} */
  const cells = [];
  cells.push(`<mxCell id="0"/>`);
  cells.push(`<mxCell id="1" parent="0"/>`);

  /** @type {Map<string,string>} */
  const nodeIds = new Map();
  let currentId = 2;

  let maxRight = 0;
  let maxBottom = 0;

  const titleWidth = Math.max(280, title.length * 8);
  cells.push(
    `<mxCell id="${currentId}" value="${xmlAttrEscape(htmlEscape(title))}" style="text;html=1;strokeColor=none;fillColor=none;align=center;verticalAlign=middle;fontSize=20;fontStyle=1;" vertex="1" parent="1">` +
      `<mxGeometry x="40" y="18" width="${titleWidth}" height="28" as="geometry"/>` +
      `</mxCell>`,
  );
  currentId++;

  const nodeStyle =
    'shape=rectangle;rounded=0;html=1;whiteSpace=wrap;overflow=fill;align=left;verticalAlign=top;fillColor=#ffffff;strokeColor=#000000;fontColor=#000000;spacing=0;';

  const sortedNodes = [...nodes.values()].sort((a, b) => a.order - b.order);
  for (const node of sortedNodes) {
    nodeIds.set(node.name, String(currentId));
    const { html, width, height } = buildNodeHtml(node);
    node.width = width;
    node.height = height;
    const x = node.x;
    const y = node.y;

    cells.push(
      `<mxCell id="${currentId}" value="${xmlAttrEscape(html)}" style="${nodeStyle}" vertex="1" parent="1">` +
        `<mxGeometry x="${x.toFixed(2)}" y="${y.toFixed(2)}" width="${width.toFixed(2)}" height="${height.toFixed(2)}" as="geometry"/>` +
        `</mxCell>`,
    );
    currentId++;
    maxRight = Math.max(maxRight, x + width);
    maxBottom = Math.max(maxBottom, y + height);
  }

  for (const rel of relations) {
    if (rel.hidden) continue;
    const srcId = nodeIds.get(rel.source);
    const dstId = nodeIds.get(rel.target);
    if (!srcId || !dstId) continue;

    const style = makeEdgeStyle(rel.token);
    // Teacher requirement: every relationship arrow must have a label.
    // Export rule: label shows ONLY the relation type (dependency/association/...).
    // Ignore any explicit label text in the source .puml to avoid showing API names.
    const inferred = inferRelationLabel(rel.token);
    const rawLabel = inferred || 'relation';
    const value = xmlAttrEscape(htmlEscape(rawLabel));
    cells.push(
      `<mxCell id="${currentId}" value="${value}" style="${style}" edge="1" parent="1" source="${srcId}" target="${dstId}">` +
        `<mxGeometry relative="1" as="geometry"/>` +
        `</mxCell>`,
    );
    currentId++;
  }

  const pageWidth = Math.max(1920, Math.floor(maxRight + 120));
  const pageHeight = Math.max(1080, Math.floor(maxBottom + 120));

  const root = `<root>${cells.join('')}</root>`;
  const mxGraph =
    `<mxGraphModel dx="1200" dy="800" grid="1" gridSize="10" guides="1" tooltips="1" connect="1" arrows="1" fold="1" page="1" pageScale="1" pageWidth="${pageWidth}" pageHeight="${pageHeight}" math="0" shadow="0">` +
    root +
    `</mxGraphModel>`;

  return mxGraph;
}

function buildMxFileSingle(pageName, mxGraphModelXml) {
  const diagramId = uuid12();
  return (
    `<?xml version="1.0" encoding="UTF-8"?>` +
    `<mxfile host="app.diagrams.net" modified="2026-03-23T00:00:00.000Z" agent="puml_to_drawio.mjs" version="26.0.11" compressed="false">` +
    `<diagram id="${diagramId}" name="${xmlAttrEscape(pageName)}">` +
    mxGraphModelXml +
    `</diagram>` +
    `</mxfile>`
  );
}

function buildMxFileMaster(pages) {
  // pages: {name, mxGraphModelXml}[]
  const diagrams = pages
    .map(
      (p) =>
        `<diagram id="${uuid12()}" name="${xmlAttrEscape(p.name)}">${p.mxGraphModelXml}</diagram>`,
    )
    .join('');
  return (
    `<?xml version="1.0" encoding="UTF-8"?>` +
    `<mxfile host="app.diagrams.net" modified="2026-03-23T00:00:00.000Z" agent="puml_to_drawio.mjs" version="26.0.11" compressed="false">` +
    diagrams +
    `</mxfile>`
  );
}

function listPumlFiles(inputPath) {
  const stat = fs.statSync(inputPath);
  if (stat.isDirectory()) {
    const out = [];
    const entries = fs.readdirSync(inputPath, { withFileTypes: true });
    for (const e of entries) {
      if (e.isFile() && e.name.toLowerCase().endsWith('.puml')) out.push(path.join(inputPath, e.name));
    }
    out.sort();
    return out;
  }
  return [inputPath];
}

function defaultOutputDir(cwd) {
  // Match the provided sample: ROOT.parent / "drawio-class-diagram"
  return path.resolve(cwd, '..', 'drawio-class-diagram');
}

function convertAll(inputs, outDir, overwrite, writeMaster, onlyCompact) {
  const pumlFiles = [];
  for (const input of inputs) {
    for (const f of listPumlFiles(input)) {
      if (f.toLowerCase().endsWith('.puml')) pumlFiles.push(f);
    }
  }
  const filtered = onlyCompact
    ? pumlFiles.filter((f) => path.basename(f).toLowerCase().endsWith('_compact.puml'))
    : pumlFiles;
  filtered.sort();

  fs.mkdirSync(outDir, { recursive: true });

  let converted = 0;
  const pages = [];

  for (const file of filtered) {
    const parsed = parsePumlFile(file);
    const mxGraph = buildDiagramXml(parsed);
    pages.push({ name: path.basename(file, '.puml'), mxGraphModelXml: mxGraph });

    const outPath = path.join(outDir, `${path.basename(file, '.puml')}.drawio`);
    if (!overwrite && fs.existsSync(outPath)) continue;
    const xml = buildMxFileSingle(path.basename(file, '.puml'), mxGraph);
    fs.writeFileSync(outPath, xml, 'utf8');
    converted++;
  }

  if (writeMaster && pages.length) {
    const masterPath = path.join(outDir, 'all-class-diagrams.drawio');
    if (overwrite || !fs.existsSync(masterPath)) {
      fs.writeFileSync(masterPath, buildMxFileMaster(pages), 'utf8');
    }
  }

  return { converted, total: filtered.length };
}

function main() {
  const args = process.argv.slice(2);
  if (!args.length) {
    console.error(
      'Usage: node puml_to_drawio.mjs <file-or-dir> [more...] [--out-dir DIR] [--overwrite] [--no-master] [--only-compact]',
    );
    process.exit(2);
  }

  let outDir = '';
  let overwrite = false;
  let writeMaster = true;
  let onlyCompact = false;
  const inputs = [];

  for (let i = 0; i < args.length; i++) {
    const a = args[i];
    if (a === '--out-dir') {
      outDir = args[i + 1] || '';
      i++;
      continue;
    }
    if (a === '--overwrite') {
      overwrite = true;
      continue;
    }
    if (a === '--no-master') {
      writeMaster = false;
      continue;
    }
    if (a === '--only-compact') {
      onlyCompact = true;
      continue;
    }
    inputs.push(a);
  }

  const resolvedOutDir = outDir ? path.resolve(outDir) : defaultOutputDir(process.cwd());
  const { converted, total } = convertAll(inputs, resolvedOutDir, overwrite, writeMaster, onlyCompact);
  console.log(`Converted ${converted} file(s) (found ${total} .puml) -> ${resolvedOutDir}`);
}

main();
